src/protocol/mod.rs

```diff
@@ -438,11 +505,16 @@ impl WebSocketContext {
             // the negotiated extensions defines the meaning of such a nonzero
             // value, the receiving endpoint MUST _Fail the WebSocket
             // Connection_.
+            let mut is_compressed = false;
             {
                 let hdr = frame.header();
-                if hdr.rsv1 || hdr.rsv2 || hdr.rsv3 {
+                if (hdr.rsv1 && self.pmce.is_none()) || hdr.rsv2 || hdr.rsv3 {
                     return Err(Error::Protocol(ProtocolError::NonZeroReservedBits));
                 }
+
+                if hdr.rsv1 && self.pmce.is_some() {
+                    is_compressed = true;
+                }
             }
```

Comment:
I think we could use something like:

```
let is_compressed = {
... <that block of code> ...
}
```

So that we don't need a `mut`.

But a general question: what if we have more options? E.g. apart from `compressed` some other extensions? In this case we'll probably need to make the implementation a bit more generic and abstract, so that we don't clutter the code. Related to the previous comment.

Reply:
Made it immutable. We can also just use `rsv1` instead of `is_compressed` if you prefer.

If we have more extensions, we'll need to first check for any conflicts in their use of `rsv` bits during handshake. Then something like this:

```rust
let (is_compressed, use_e2, use_e3) = self.validate_header(frame.header())?;
// or
let (is_compressed, use_e2, use_e3) = {
    let hdr = frame.header();
    if (hdr.rsv1 && self.pmce.is_none()) || (hdr.rsv2 && self.ext2.is_none()) || (hdr.rsv3 && self.ext3.is_none()) {
        return Err(Error::Protocol(ProtocolError::NonZeroReservedBits));
    }
    (hdr.rsv1, hdr.rsv2, hdr.rsv3)
};
```

I don't know if an extension can use more than one bit.